universe/Condition.cpp

```diff
@@ -9955,6 +9955,203 @@ unsigned int Not::GetCheckSum() const {
+    ConditionBase(),
+    m_operands(std::move(operands))
+{}
+
+bool TopmostMatches::operator==(const ConditionBase& rhs) const {
+    if (this == &rhs)
+        return true;
+    if (typeid(*this) != typeid(rhs))
+        return false;
+
+    const TopmostMatches& rhs_ = static_cast<const TopmostMatches&>(rhs);
+
+    if (m_operands.size() != rhs_.m_operands.size())
+        return false;
+    for (unsigned int i = 0; i < m_operands.size(); ++i) {
+        CHECK_COND_VREF_MEMBER(m_operands.at(i))
+    }
+
+    return true;
+}
+
+void TopmostMatches::Eval(const ScriptingContext& parent_context, ObjectSet& matches,
+               ObjectSet& non_matches, SearchDomain search_domain/* = NON_MATCHES*/) const
+{
+    std::shared_ptr<const UniverseObject> no_object;
+    ScriptingContext local_context(parent_context, no_object);
+
+    if (m_operands.empty()) {
+        ErrorLogger() << "TopmostMatches::Eval given no operands!";
+        return;
+    }
+    for (auto& operand : m_operands) {
+        if (!operand) {
+            ErrorLogger() << "TopmostMatches::Eval given null operand!";
+            return;
+        }
+    }
+
+    // probably i need to figure out which condition matches first
+    // by searching both domains
+    // ... or actually i can propagate the search domain and look for the first match, probably i can
+    // Expressions like  'NOT TopmostMatches [ A B C ]' matches all candidates which do not match the topmost matching condition
+    // which means the same as TopmostMatches [ (Not A) (Not B) (Not C) ]
+    // TODO use logic like in base eval for both domains
+    if (search_domain == NON_MATCHES) {
+        // check all operand conditions on all objects in the non matches set
+        // stop if the operand condition has at least one non match
+        ObjectSet temp_matches;
+        temp_matches.reserve(non_matches.size());
+        // try matching operand conditions until one matches
+        for (auto& operand : m_operands) {
+            // move back temp_matches
+            non_matches.insert(non_matches.end(),
+                           std::make_move_iterator(temp_matches.begin()),
+                           std::make_move_iterator(temp_matches.end()));
+            temp_matches.clear();
+            operand->Eval(local_context, temp_matches, non_matches, NON_MATCHES);
+            if (!non_matches.empty()) break;
+        }
+
+        // operand matching items were already moved from non_matches to temp_matches
+        // move operand matching items from temp_matches to matches
+        matches.insert(matches.end(),
+                       std::make_move_iterator(temp_matches.begin()),
+                       std::make_move_iterator(temp_matches.end()));
+    } else /*(search_domain == MATCHES)*/ {
+        // check all operand conditions on all objects in the matches set
+        // stop if the operand condition matches at least one item
```

Comment:
Don't think that makes sense... You'd need to check if anything in the matches or non_matches is matched by the current subcondition to decide which subcondition to use, then go back and move stuff out of matches that doesn't match the selected subcondition. You can't just test the matches to pick which subcondition to apply to matches.


Not sure that makes sense. This condition doesn't match objects independent of all the others,

Reply:
I tried to make the documentation better for the MATCHES implementation

Reply:
NON_MATCHES and MATCHES domain work now